## 당뇨병 여부 판단
- 이상치 처리 (Glucose, BloodPressure, SkinThickness, Insulin, BMI가 0인 값)

## [참고]작업형2 문구
- 출력을 원하실 경우 print() 함수 활용
- 예시) print(df.head())
- getcwd(), chdir() 등 작업 폴더 설정 불필요
- 파일 경로 상 내부 드라이브 경로(C: 등) 접근 불가

### 데이터 파일 읽기 예제
- import pandas as pd
- X_test = pd.read_csv("data/X_test.csv")
- X_train = pd.read_csv("data/X_train.csv")
- y_train = pd.read_csv("data/y_train.csv")

### 사용자 코딩

### 답안 제출 참고
- 아래 코드 예측변수와 수험번호를 개인별로 변경하여 활용
- pd.DataFrame({'cust_id': X_test.cust_id, 'gender': pred}).to_csv('003000000.csv', index=False)

In [ ]:
# 시험환경 세팅 (코드 변경 X)
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

def exam_data_load(df, target, id_name="", null_name=""):
    if id_name == "":
        df = df.reset_index().rename(columns={"index": "id"})
        id_name = 'id'
    else:
        id_name = id_name
    
    if null_name != "":
        df[df == null_name] = np.nan
    
    X_train, X_test = train_test_split(df, test_size=0.2, random_state=2021)
    
    y_train = X_train[[id_name, target]]
    X_train = X_train.drop(columns=[target])

    
    y_test = X_test[[id_name, target]]
    X_test = X_test.drop(columns=[target])
    return X_train, X_test, y_train, y_test 
    
df = pd.read_csv("../input/pima-indians-diabetes-database/diabetes.csv")
X_train, X_test, y_train, y_test = exam_data_load(df, target='Outcome')

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((614, 9), (154, 9), (614, 2), (154, 2))

<hr style="border: solid 3px blue;">


![](https://media1.giphy.com/media/tKxL0z3hHVJ8InrFid/giphy.gif)

Picture Credit: https://media1.giphy.com

깡초보인 석규는 걱정이 많습니다. 판다스 및 sklearn등에 대해서 아직 자신이 없습니다.
하지만, 시험에 합격을 하고 싶은 마음은 간절 합니다.

이제부터 시험 시작 입니다. 떨리는 마음을 달래며 시험을 보도록 하겠습니다.

석규는 **작업형2**을 먼저 보기로 했습니다. 

왜냐하면, **작업형2**의 배점이 40점이라서 여기서 망치면 거의 불합격이기 때문입니다.

----------------------------------------------------------------------
# Start

![](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FofRef%2Fbtq7peXUL1f%2FXqbL4h7oK7crd19nJjC6Uk%2Fimg.png)

Picture Credit: https://img1.daumcdn.net


위와 같은 환경에서 진행을 합니다. 시각화는 안 되고 print 문을 잘 사용해야 합니다.

또한, dir()과 help()을 적절히 잘 이용하면 중간에 기억이 나지 않을 때 도움이 많이 됩니다.

------------------------------------------------------------------------------------------------

## 라이브러리 및 데이터 불러오기

시험을 위한 package들을 import 합니다.

석규는 앙상블 방법 중에 RondomForest classifier을 사용하기로 했습니다.

빅분기 수준의 문제는 RondomForest classifier을 사용하는 것이 적절할 수 있습니다.

분류 문제가 나온다면 따지지 말고 RondomForest classifier을 이용하는 것이 좋을 것 같고, 이것만 잘 연습해 가도 되지 않을까라는 생각이 듭니다.

아래와 같이 외우기 좋게 일부러 문장을 만들었습니다.

**석규는 앙상블(ensemble)을 배운다.**
* **sk learn ensemble**

In [2]:
# 라이브러리
import pandas as pd
import sklearn

일단 어렵게 위와 같이 import을 해 놓았습니다. 그런데 갑자기 sklearn이 어떤 라이브러리를 가지고 있는지 기억이 안 나네요.

그런 경우는 help을 이용합니다.

In [3]:
help(sklearn)

Help on package sklearn:

NAME
    sklearn

DESCRIPTION
    Machine learning module for Python
    
    sklearn is a Python module integrating classical machine
    learning algorithms in the tightly-knit world of scientific Python
    packages (numpy, scipy, matplotlib).
    
    It aims to provide simple and efficient solutions to learning problems
    that are accessible to everybody and reusable in various contexts:
    machine-learning as a versatile tool for science and engineering.
    
    See http://scikit-learn.org for complete documentation.

PACKAGE CONTENTS
    __check_build (package)
    _build_utils (package)
    _config
    _distributor_init
    _isotonic
    _loss (package)
    base
    calibration
    cluster (package)
    compose (package)
    conftest
    covariance (package)
    cross_decomposition (package)
    datasets (package)
    decomposition (package)
    discriminant_analysis
    dummy
    ensemble (package)
    exceptions
    experimental (package)
    ext

오! **PACKAGE CONTENTS**에 'ensemble'이 있네요. 그런데, 외운 것이 잘 기억이 안 나네요.
Random.. 뭐라고 하는 것이 있었는데 말이죠.
이때 dir을 사용합니다.

In [4]:
import sklearn.ensemble
dir(sklearn.ensemble)

['AdaBoostClassifier',
 'AdaBoostRegressor',
 'BaggingClassifier',
 'BaggingRegressor',
 'BaseEnsemble',
 'ExtraTreesClassifier',
 'ExtraTreesRegressor',
 'GradientBoostingClassifier',
 'GradientBoostingRegressor',
 'IsolationForest',
 'RandomForestClassifier',
 'RandomForestRegressor',
 'RandomTreesEmbedding',
 'StackingClassifier',
 'StackingRegressor',
 'VotingClassifier',
 'VotingRegressor',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_bagging',
 '_base',
 '_forest',
 '_gb',
 '_gb_losses',
 '_gradient_boosting',
 '_iforest',
 '_stacking',
 '_voting',
 '_weight_boosting',
 'typing']

아 'RandomForestClassifier'가 있네요.

참, 문제를 읽을 때 해당 문제가 분류 문제인지 회귀 문제인지 먼저 판단해야 하는데, 이 경우는 분류 문제이니까, Classifier로 끝나는 라이브러리를 import 해야 하겠죠.

In [5]:
from sklearn.ensemble import RandomForestClassifier

이제 **RandomForestClassifier**을 잘 import 했네요.

----------------------------------------------------------------
# EDA

![](https://media3.giphy.com/media/l3vRiLydJmh7X0Yx2/giphy.gif?cid=c9f270ba5a2109a369344d76635aea76)

Picture Credit: https://media3.giphy.com

자, 이제 EDA(Exploratory data analysis)을 해야죠.
EDA을 할려고 하니, **머리(HEAD=>HIDI)**가 아프네요.

우리 **판다들(Pandas)**도 머리가 아픈가 봐요.

HIDI는 다음과 같습니다.
* H => head()
* I => info()
* D => describe()
* I => isnull()


피마 인디언 당뇨병에 대한 EDA에 대해서 좀 더 알고 싶다면, 아래 notebook을 참고하세요.

https://www.kaggle.com/ohseokkim/diabetes-three-ensemble-models

In [6]:
X_train.head().T

,147,344,390,150,132
id,147.0,344.000,390.000,150.000,132.000
Pregnancies,2.0,8.000,1.000,1.000,3.000
Glucose,106.0,95.000,100.000,136.000,170.000
BloodPressure,64.0,72.000,66.000,74.000,64.000
SkinThickness,35.0,0.000,29.000,50.000,37.000
Insulin,119.0,0.000,196.000,204.000,225.000
BMI,30.5,36.800,32.000,37.400,34.500
DiabetesPedigreeFunction,1.4,0.485,0.444,0.399,0.356
Age,34.0,57.000,42.000,24.000,30.000


한가지 tip은 head().T을 사용해서 열들이 위와 같이 행처럼 보이게 하는 것이 열들을 파악하는데 좋을 수 있어요.

여기서 어떤 열들이 있는지 전체적으로 파악을 해 보는 것이 좋을 것 같네요.

In [7]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 614 entries, 147 to 116
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        614 non-null    int64  
 1   Pregnancies               614 non-null    int64  
 2   Glucose                   614 non-null    int64  
 3   BloodPressure             614 non-null    int64  
 4   SkinThickness             614 non-null    int64  
 5   Insulin                   614 non-null    int64  
 6   BMI                       614 non-null    float64
 7   DiabetesPedigreeFunction  614 non-null    float64
 8   Age                       614 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 48.0 KB


여기서는 각 열들의 속성을 파악해 보아요. Dtype이 object인 경우에는 인코딩을 고려해야 하지만, 다행히 모두 수치형이네요.

한가지 더 확인할 내용은 결측치가 있는지 입니다. 다행이 이 경우에는 결측치가 없네요.

In [8]:
X_train.describe()

,id,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
count,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000,614.000000
mean,386.978827,3.845277,120.705212,68.530945,20.982085,82.229642,31.948534,0.477471,33.319218
std,219.881344,3.388952,32.047977,19.761915,15.965031,116.525772,7.734385,0.339811,11.940953
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000
25%,196.500000,1.000000,99.000000,62.000000,0.000000,0.000000,27.400000,0.244250,24.000000
50%,388.000000,3.000000,117.000000,72.000000,23.000000,44.000000,32.250000,0.377000,29.000000
75%,575.750000,6.000000,140.000000,80.000000,33.000000,130.000000,36.600000,0.628500,41.000000
max,766.000000,17.000000,199.000000,114.000000,99.000000,846.000000,59.400000,2.420000,81.000000


눈에 띄는 이상치가 있는지 대충 확인해 봐요. 이 경우에는 최소값으로 0으로 보이는 열들이 많네요. 

이상치 탐지 및 처리는 다소 까다로워서 실제 시험에 이상치를 처리해야 하는 문제가 나올지 모르겠네요.

In [9]:
X_train.isnull().sum()

id                          0
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
dtype: int64

In [10]:
X_test.isnull().sum()

id                          0
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
dtype: int64

isnull().sum()을 통해서 열별로 결측치가 있는지 확인해 보았는데, 다행히 train/test dataset 모두 결측치가 없네요.

------------------------------------------------------------------------------------------------
## 데이터 전처리

저는 빅분기 수준의 데이터 전처리는 결측치 처리와 인코딩과 feature selection 하는 정도라고 생각합니다.

더이상 잘 할려고 하는 것은 **미친(MAD => MED)** 짓이죠.

**MED**
* M -> 결측치 처리, fillna() 면 충분할 것 같습니다. 이것은 다른 notebook에서 언급해야 할 것 같네요.
* E -> Encoding, 판다스의 get_dummies()면 충분할 것 같습니다. 이것도 다른 notebook에서 언급해야 할 것 같네요.
* D -> Drop

인코딩에 대해서 좀 더 알고 싶다면, 아래 notebook을 참고 하세요.

https://www.kaggle.com/ohseokkim/preprocessing-encoding-categorical-data

여기서는 drop만 합니다. 그런데 drop()을 쓸려고 하는 막막하네요.

**지금 이 시점에서 도움이 필요합니다.**
* **help**

In [11]:
help(df.drop)

Help on method drop in module pandas.core.frame:

drop(labels=None, axis: 'Axis' = 0, index=None, columns=None, level: 'Level | None' = None, inplace: 'bool' = False, errors: 'str' = 'raise') method of pandas.core.frame.DataFrame instance
    Drop specified labels from rows or columns.
    
    Remove rows or columns by specifying label names and corresponding
    axis, or by specifying directly index or column names. When using a
    multi-index, labels on different levels can be removed by specifying
    the level. See the `user guide <advanced.shown_levels>`
    for more information about the now unused levels.
    
    Parameters
    ----------
    labels : single label or list-like
        Index or column labels to drop.
    axis : {0 or 'index', 1 or 'columns'}, default 0
        Whether to drop labels from the index (0 or 'index') or
        columns (1 or 'columns').
    index : single label or list-like
        Alternative to specifying axis (``labels, axis=0``
        is equiv

오! 파라미터와 리턴과 실제 예들까지 있네요. 좋습니다. 이걸로 drop()을 구현해 보죠.
특히, 세개의 parameter들을 많이 이용할 것 같네요.

    labels : single label or list-like
        Index or column labels to drop.
        
    axis : {0 or 'index', 1 or 'columns'}, default 0
        Whether to drop labels from the index (0 or 'index') or
        columns (1 or 'columns').
        
    inplace : bool, default False
        If False, return a copy. Otherwise, do operation
        inplace and return None.        
        

일반적으로 id열들은 학습에 전혀 도움이 되지 않습니다. 아니 어떤 경우에는 id로 인해서 오퍼피팅이 되어서 테스트 결과가 안 좋을 수 있습니다.
ID 계열의 열들은 제거하는 것이 좋습니다.

In [12]:
# id 제외
X_train.drop('id',axis=1,inplace=True)
X_test.drop('id',axis=1,inplace=True)

잘 제거가 되었는지 확인해 보죠.

In [13]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 614 entries, 147 to 116
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               614 non-null    int64  
 1   Glucose                   614 non-null    int64  
 2   BloodPressure             614 non-null    int64  
 3   SkinThickness             614 non-null    int64  
 4   Insulin                   614 non-null    int64  
 5   BMI                       614 non-null    float64
 6   DiabetesPedigreeFunction  614 non-null    float64
 7   Age                       614 non-null    int64  
dtypes: float64(2), int64(6)
memory usage: 43.2 KB


잘 제거가 되었네요.

----------------------------------------------------
# 학습 및 예측

![](https://t1.daumcdn.net/cfile/tistory/23649E4255FA72C71D)

Picture Credit: https://t1.daumcdn.net

거의 마지막까지 다 왔습니다. **부디 모두들 목표를 이루시고 살아오시기를....**

이 시점에서 다시 한번 help을 외칩니다. RandomForestClassifier가 뭐지?

In [14]:
help(RandomForestClassifier)

Help on class RandomForestClassifier in module sklearn.ensemble._forest:

class RandomForestClassifier(ForestClassifier)
 |  RandomForestClassifier(n_estimators=100, *, criterion='gini', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features='auto', max_leaf_nodes=None, min_impurity_decrease=0.0, min_impurity_split=None, bootstrap=True, oob_score=False, n_jobs=None, random_state=None, verbose=0, warm_start=False, class_weight=None, ccp_alpha=0.0, max_samples=None)
 |  
 |  A random forest classifier.
 |  
 |  A random forest is a meta estimator that fits a number of decision tree
 |  classifiers on various sub-samples of the dataset and uses averaging to
 |  improve the predictive accuracy and control over-fitting.
 |  The sub-sample size is controlled with the `max_samples` parameter if
 |  `bootstrap=True` (default), otherwise the whole dataset is used to build
 |  each tree.
 |  
 |  Read more in the :ref:`User Guide <forest>`.
 |  
 |  P

엄청 자세히 나왔네요.그리고 아래와 같은 example도 있네요. ㅋㅋㅋ

>  Examples
>    --------
>    from sklearn.ensemble import RandomForestClassifier
> 
>    from sklearn.datasets import make_classification
> 
>    X, y = make_classification(n_samples=1000, n_features=4,
>    ...                            n_informative=2, n_redundant=0,
>    ...                            random_state=0, shuffle=False)
> 
>    clf = RandomForestClassifier(max_depth=2, random_state=0)
> 
>    clf.fit(X, y)
> 
>    RandomForestClassifier(...)
> 
>    print(clf.predict([[0, 0, 0, 0]]))

이걸로 훈련 및 학습을 해 보죠.



In [15]:
model = RandomForestClassifier() 
model.fit(X_train, y_train['Outcome'])
predictions = model.predict(X_test)

In [16]:
# 오버피팅 되었을 경우에 점수가 잘나올 수 있음 (객관적인 평가 아님, 밸리데이션 데이터로 평가 필요함)
round(model.score(X_train, y_train['Outcome']) * 100, 2) 

100.0

Wow! 우리 model이 완전히 답을 외웠네요. 이런 경우는 오버피팅이 되었다고 생각을 해도 될 것 같네요.
하지만, 빅분기 실기에서는 오버피팅도 큰 문제가 되지 않을 문제들이 나오지 않을까라는 생각이 듭니다.

In [17]:
output = pd.DataFrame({'idx': X_test.index, 'Outcome': predictions})
output.head()

,idx,Outcome
0,258,1
1,220,1
2,438,0
3,130,1
4,730,0


In [18]:
# 수험번호.csv로 출력
output.to_csv('1234567.csv', index=False)

![](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FofRef%2Fbtq7peXUL1f%2FXqbL4h7oK7crd19nJjC6Uk%2Fimg.png)

Picture Credit: https://img1.daumcdn.net

마지막으로 제출 버튼을 누르는 것 잊지 마세요.

--------------------------------------------------------------------------------------------------
## 결과 채점 (수험자는 알 수 없는 부분임)

In [19]:

round(model.score(X_test, y_test['Outcome']) * 100, 2) 

77.27

77.27점이 나왔네요. 이 정도면 빅분기 실기 시험에서 어느정도의 점수는 확보 되었다고 봐도 될 것 같습니다.
다른 문제로 넘어가도 될 것 같네요.

![](https://reactiongifs.me/wp-content/uploads/2019/10/Well-Done.gif)

Picture Credit: https://reactiongifs.me

<hr style="border: solid 3px blue;">
